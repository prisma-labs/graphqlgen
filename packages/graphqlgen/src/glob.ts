import * as glob from 'glob'
import { File } from 'graphqlgen-json-schema'
import { getPath } from './parse'

/**
 * Returns the path array from glob patterns
 */
export const extractGlobPattern = (file: File) => {
  return glob.sync(getPath(file))
}

/**
 * Handles the glob pattern of models.files
 */
export const handleGlobPattern = (files?: File[]): File[] => {
  if (!files) {
    return []
  }

  return files.reduce<File[]>((acc, file) => {
    const globedPaths = extractGlobPattern(file)

    if (globedPaths.length === 0) {
      return [...acc, file]
    }

    const globedFiles: File[] = globedPaths.map(path => {
      if (typeof file === 'string') {
        return path
      }

      return {
        path,
        defaultName: file.defaultName,
      }
    })

    return [...acc, ...globedFiles]
  }, [])
}
