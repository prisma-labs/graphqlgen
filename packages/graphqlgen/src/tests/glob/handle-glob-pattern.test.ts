import { handleGlobPattern } from '../../glob'

/**
 * Test the glob pattern handling of Files[] : Basic
 */
test('Handle Basic Walk', () => {
  expect(
    handleGlobPattern([
      { path: 'anotherPath', defaultName: undefined },
      './src/tests/fixtures/glob/dir-1/*.ts',
      { path: 'somepath', defaultName: undefined },
    ]),
  ).toMatchSnapshot()
})

/**
 * Test the glob pattern handling of Files[] : Complex
 */
test('Handle Complex Walk', () => {
  expect(
    handleGlobPattern([
      { path: 'somepath', defaultName: undefined },
      './src/tests/fixtures/glob/**/*.ts',
      { path: 'anotherPath', defaultName: undefined },
    ]),
  ).toMatchSnapshot()
})
