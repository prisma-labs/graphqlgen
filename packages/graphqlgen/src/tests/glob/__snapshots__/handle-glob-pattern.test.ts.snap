// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Handle Basic Walk 1`] = `
Array [
  Object {
    "defaultName": undefined,
    "path": "anotherPath",
  },
  "./src/tests/fixtures/glob/dir-1/file-11.ts",
  "./src/tests/fixtures/glob/dir-1/file-12.ts",
  Object {
    "defaultName": undefined,
    "path": "somepath",
  },
]
`;

exports[`Handle Complex Walk 1`] = `
Array [
  Object {
    "defaultName": undefined,
    "path": "somepath",
  },
  "./src/tests/fixtures/glob/dir-1/file-11.ts",
  "./src/tests/fixtures/glob/dir-1/file-12.ts",
  "./src/tests/fixtures/glob/dir-2/file-21.ts",
  "./src/tests/fixtures/glob/dir-2/file-22.ts",
  Object {
    "defaultName": undefined,
    "path": "anotherPath",
  },
]
`;
